homeassistant/components/homekit_controller/__init__.py

```diff
@@ -133,10 +133,31 @@ def device_config_callback(self, callback_data):
         import homekit
         pairing_id = str(uuid.uuid4())
         code = callback_data.get('code').strip()
-        self.pairing_data = homekit.perform_pair_setup(
-            self.conn, code, pairing_id)
+        try:
+            self.pairing_data = homekit.perform_pair_setup(self.conn, code,
+                                                           pairing_id)
+        except homekit.exception.UnavailableError:
+            error_msg = "This accessory is already paired to another device. \
+                         Please reset the accessory and try again."
+            _configurator = self.hass.data[DOMAIN+self.hkid]
+            self.configurator.notify_errors(_configurator, error_msg)
+            return
+        except homekit.exception.AuthenticationError:
+            error_msg = "Incorrect HomeKit code for {}. Please check it and \
+                         try again.".format(self.model)
+            _configurator = self.hass.data[DOMAIN+self.hkid]
+            self.configurator.notify_errors(_configurator, error_msg)
+            return
+        except homekit.exception.UnknownError:
+            error_msg = "Received an unknown error. Please file a bug."
+            _configurator = self.hass.data[DOMAIN+self.hkid]
+            self.configurator.notify_errors(_configurator, error_msg)
```

Comment:
If it's an unknown error, wouldn't it be good to print the stack trace? If I understand correctly, when an `UnknownError` is thrown the only message that's displayed is "Received an unknown error. [...]" and debugging an issue is going to be difficult with just that error message.

Maybe replace the `return` on the line below with`raise` so that a) the configurator still shows an error but b) the complete stack trace is shown?

Reply:
That seems reasonable. The problem with UnknownError is that the device literally returned what the HomeKit spec defines as an unknown error, so it's unlikely that we're going to get much more from the stack trace, but it's going to be better than nothing.